### PR TITLE
KFSPTS-30235 Create a Notes text drop-down

### PR DESCRIPTION
**Pull request by db483**

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/module/purap/businessobject/IWantNote.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package edu.cornell.kfs.module.purap.businessobject;
+
+import org.kuali.kfs.krad.bo.PersistableBusinessObjectBase;
+
+public class IWantNote extends PersistableBusinessObjectBase {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/businessobject/IWantNote.java:5]: similar thought I had on the the SQL PR, should this be named something like IWantNoteDefault or something so it doesn't like this is the actual note text on the Iwant Document itself?

File: src/main/java/edu/cornell/kfs/module/purap/businessobject/options/IWantNoteComparator.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package edu.cornell.kfs.module.purap.businessobject.options;
+
+import java.util.Comparator;
+
+import edu.cornell.kfs.module.purap.businessobject.IWantNote;
+
+public class IWantNoteComparator implements Comparator {
+
+    @Override
+    public int compare(final Object o1, final Object o2) {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/businessobject/options/IWantNoteComparator.java:10]: Is this needed if the BO IWantNote is updated to implement comparable?

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java [hunk 3]
```diff
@@ -904,15 +904,21 @@ public ActionForward insertBONote(ActionMapping mapping, ActionForm form, HttpSe
         }
         
         // If the note text is blank, set the attachment description as the text. Otherwise, concatenate both to form the text.
+        if(StringUtils.isBlank(iWantDocumentForm.getIWantDocument().getNoteTextOption())) {
         if (StringUtils.isBlank(note.getNoteText())) {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java:908]: Does this need to be indented?

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/IWantNoteRule.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+public class IWantNoteRule extends MaintenanceDocumentRuleBase {
+
+    @Override
+    protected boolean processCustomRouteDocumentBusinessRules(MaintenanceDocument document) {
+        boolean isValid = super.processCustomRouteDocumentBusinessRules(document);
+        if (isValid) {
+            final IWantNote iWantNote = (IWantNote) document.getNewMaintainableObject().getBusinessObject();
+            if (ObjectUtils.isNull(iWantNote.getNoteIdentifier())) {
+                final SequenceAccessorService sequenceAccessorService = SpringContext
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/document/validation/IWantNoteRule.java:20]: Is SequenceAccessorService needed?  If the note Identifier isn't set, the business object service should get the next sequence from OJB.

**Review by jhulslander — COMMENTED** [review]
Changes look good for the most part, I found a few things to consider.

File: src/main/java/edu/cornell/kfs/module/purap/businessobject/IWantNote.java (hunk 1 above)
Comment (db483) [reply src/main/java/edu/cornell/kfs/module/purap/businessobject/IWantNote.java:5]: Good point. How about IWantNoteDropdownValue ?

File: src/main/java/edu/cornell/kfs/module/purap/businessobject/options/IWantNoteComparator.java (hunk 2 above)
Comment (db483) [reply src/main/java/edu/cornell/kfs/module/purap/businessobject/options/IWantNoteComparator.java:10]: Good point. I was basing this on the AccountTypeValuesFinder example.

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/IWantNoteRule.java (hunk 4 above)
Comment (db483) [reply src/main/java/edu/cornell/kfs/module/purap/document/validation/IWantNoteRule.java:20]: Great question. The issue I was trying to solve is the fact that the maintenance doc will not have the note identifier if the value is set through the OJB. The new BO will have the identifier set and saved in the DB but the Doc does not have the value. I tried a couple other ways to try to force the value on the doc but this was the only way I got it to work (and the only example I found in the code base). If you have another idea I could try please let me know. Thank you!

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java (hunk 3 above)
Comment (db483) [reply src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java:908]: updated

File: src/main/java/edu/cornell/kfs/module/purap/document/validation/IWantNoteRule.java (hunk 4 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/document/validation/IWantNoteRule.java:20]: Thanks for the explanation, I think this solution makes sense.

File: src/main/resources/edu/cornell/kfs/module/purap/cu-ojb-purap.xml [hunk 5]
```diff
@@ -334,6 +334,14 @@
 	    </collection-descriptor>-->
 	</class-descriptor>
     
+    <class-descriptor class="edu.cornell.kfs.module.purap.businessobject.IWantNoteDropdownValue" table="CU_PUR_IWNT_NTE_DRPDN_VAL_T">
+        <field-descriptor name="noteIdentifier" column="NTE_ID" jdbc-type="INTEGER" primarykey="true" index="true" autoincrement="true" sequence-name="CU_PUR_IWNT_NTE_TXT_ID_SEQ"/>
+        <field-descriptor name="objectId" column="OBJ_ID" jdbc-type="VARCHAR" index="true"/>
+        <field-descriptor name="versionNumber" column="VER_NBR" jdbc-type="BIGINT" locking="true"/>
+        <field-descriptor name="noteText" column="NTE_TXT" jdbc-type="VARCHAR" index="true"/>
+        <field-descriptor name="active" column="ROW_ACTV_IND" jdbc-type="VARCHAR" index="true" conversion="org.kuali.kfs.core.framework.persistence.ojb.conversion.OjbCharBooleanConversion"/>
+    </class-descriptor>
```
Comment (jhulslander) [inline src/main/resources/edu/cornell/kfs/module/purap/cu-ojb-purap.xml:343]: should the last update time stamp be added?

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java [hunk 6]
```diff
@@ -904,15 +904,21 @@ public ActionForward insertBONote(ActionMapping mapping, ActionForm form, HttpSe
         }
         
         // If the note text is blank, set the attachment description as the text. Otherwise, concatenate both to form the text.
-        if (StringUtils.isBlank(note.getNoteText())) {
-            note.setNoteText(iWantDocumentForm.getIWantDocument().getAttachmentDescription());
-        } else {
-            note.setNoteText(iWantDocumentForm.getIWantDocument().getAttachmentDescription() + ": " + note.getNoteText());
+        if(StringUtils.isBlank(iWantDocumentForm.getIWantDocument().getNoteTextOption())) {
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java:907]: the formatting looks off in this block. A space  is missing between the if the open paren, and the else on line 914 could be on the line above

**Review by jhulslander — CHANGES_REQUESTED** [review]
changes look good, I found a few things to consider

File: src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java (hunk 6 above)
Comment (db483) [reply src/main/java/edu/cornell/kfs/module/purap/document/web/struts/IWantDocumentAction.java:907]: updated

File: src/main/resources/edu/cornell/kfs/module/purap/cu-ojb-purap.xml (hunk 5 above)
Comment (db483) [reply src/main/resources/edu/cornell/kfs/module/purap/cu-ojb-purap.xml:343]: looks like I didn't commit this change

**Review by jhulslander — APPROVED** [review]
The code looks good and can be merged when the SQL is ready

**Review by jhulslander — APPROVED** [review]
code changes look good, unit tests pas and SQL looks good, I shall merge

**Pull request merged**